aten/src/ATen/native/cuda/TensorShape.cu

```diff
@@ -268,6 +281,301 @@ std::pair<at::Tensor, std::vector<int64_t*>> pack_vecs(
+// Assume elements of src and dst have the same data type.
+template <>
+__device__ __inline__ void copy_chunk_with_pad<void, void>(
+    char* dst,
+    const char* src,
+    int64_t max_chunk_size,
+    int64_t actual_chunk_size,
+    int64_t thread_idx,
+    int64_t num_threads) {
+  // Fast path when the number of threads is larger than the number of bytes to
+  // be copied (i.e., max_chunk_size). In this case, each thread only copies 1
+  // byte. For 0 <= thread_idx < actual_chunk_size, the thread copies data from
+  // `src`. For actual_chunk_size <= thread_idx < max_chunk_size, the thread set
+  // the val=0 for padding.
+  if (max_chunk_size < num_threads) {
+    int64_t val = 0;
+    if (thread_idx < actual_chunk_size) {
+      val = src[thread_idx];
+    }
+    if (thread_idx < max_chunk_size) {
+      dst[thread_idx] = val;
+    }
+    return;
+  }
+  // Split dst array into three parts:
+  // [dst, dst+align_off), [dst+align_off, dst+align_end), [dst+align_end,
+  // dst+max_chunk_size) The second part is aligned with BYTES_PER_THREAD(=16
+  // bytes) to enable `stream_store128`.
+  int64_t align_off, aligned_size;
+  get_aligned_region(
+      dst, actual_chunk_size, BYTES_PER_THREAD, align_off, aligned_size);
+  int64_t align_end = align_off + aligned_size;
+  for (int64_t i = align_off + thread_idx * BYTES_PER_THREAD; i < align_end;
+       i += num_threads * BYTES_PER_THREAD) {
+    uint4 val;
+    if (is_aligned<uint4>(src + i)) {
+      stream_load128(val, src + i);
+    } else {
+      for (size_t j = 0; j < BYTES_PER_THREAD; ++j) {
+        reinterpret_cast<char*>(&val)[j] = src[i + j];
+      }
+    }
+    stream_store128(&dst[i], val);
+  }
+  // Copy data for the first part of dst array [dst, dst+align_off).
+  // Check `thread_idx<max_chunk_sze` for the edge case that max_chunk_size <
+  // align_off.
+  if (thread_idx < align_off && thread_idx < max_chunk_size) {
+    char val = (char)0;
+    if (thread_idx < actual_chunk_size) {
+      val = src[thread_idx];
+    }
+    dst[thread_idx] = val;
+  }
+  // Copy data for the third part of dst array [dst+align_end,
+  // dst+max_chunk_size).
+  while (align_end + thread_idx < max_chunk_size) {
+    char val = (char)0;
+    if (align_end + thread_idx < actual_chunk_size) {
+      val = src[align_end + thread_idx];
+    }
+    dst[align_end + thread_idx] = val;
+    align_end += num_threads;
+  }
+}
+
+// Read `actual_chunk_size` bytes from bfloat16 array `src`  (i.e.,
+// actual_chunk_size/2 bfloat16 elements), cast bfloat16 to float, and write to
+// float array `dst`. Pad 0.0f if 2*actual_chunk_size < max_chunk_size.
+// src is pointer to a bfloat16 array; dst is pointer to a float32 array;
+// max_chunk_size is number of bytes in dst array; actual_chunk_size is
+// the number of bytes in src array, which is less than or equal to
+// max_chunk_size / 2.
+template <>
+__device__ __inline__ void copy_chunk_with_pad<float, bfloat16>(
```

Comment:
I don't think explicit use of the intrinic and the explicit template instantiation here is necessary: https://github.com/pytorch/pytorch/blob/main/c10/util/BFloat16-inl.h#L41